### PR TITLE
fix(YouTube - SponsorBlock): Skip segments when casting

### PR DESCRIPTION
**Pull request by thedroidgeek**
Add patch code to support skipping segments while casted to a big screen device, by hooking the MDX director class to intercept object instances in order to be able to call the seekTo method, similar to how it's done in the 'regular' player class.

Fixes https://github.com/ReVanced/revanced-patches/issues/374.

Integrations PR: https://github.com/ReVanced/revanced-integrations/pull/655

### CODE REVIEW

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 1]
```diff
@@ -65,34 +74,34 @@ object VideoInformationPatch : BytecodePatch(
...
         }
 
+        with(MdxPlayerDirectorFingerprint.result!!) {
+            mdxInitMethod = mutableClass.methods.first { MethodUtil.isConstructor(it) }
+
+            // find the location of the first invoke-direct call and extract the register storing the 'this' object reference
+            mdxInitInsertIndex = mdxInitMethod.implementation!!.instructions.indexOfFirst {
+                it.opcode == Opcode.INVOKE_DIRECT && ((it as Instruction35c).reference as MethodReference).name == "<init>"
+            } + 1
+            mdxInitInsertRegister = mdxInitMethod.getInstruction<Instruction35c>(mdxInitInsertIndex - 1).registerC
```
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:90]: I kept the static values for easy maintenance - the dynamic resolution code was added after I realized the method was updated and the values had to change between 19.11.43 and 19.16.39.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt [hunk 2]
```diff
@@ -0,0 +1,8 @@
+package app.revanced.patches.youtube.video.information.fingerprints
+
+
+import app.revanced.patcher.fingerprint.MethodFingerprint
+
+internal object MdxPlayerDirectorFingerprint : MethodFingerprint(
+    strings = listOf("MdxDirector setVideoStage ad should be null when videoStage is not an Ad state ")
```
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt:7]: The code was successfully tested on 18.48.39, 19.04.37, 19.11.43 and 19.16.39 - will test on more versions if needed.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt [hunk 3]
```diff
@@ -0,0 +1,14 @@
...
+import app.revanced.patcher.extensions.or
+import app.revanced.patcher.fingerprint.MethodFingerprint
+import com.android.tools.smali.dexlib2.AccessFlags
+
+internal object MdxSeekFingerprint : MethodFingerprint(
+    accessFlags = AccessFlags.PUBLIC or AccessFlags.FINAL,
+    returnType = "Z",
+    parameters = listOf("J", "L"),
+    customFingerprint = { methodDef, _ ->
+        methodDef.implementation!!.instructions.count() == 3
```
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt:12]: There's actually a very similar method (seems to have the behavior of seekToRelative) with the exact same signature (arguments and return type), with exactly the same opcodes as the one I hook (except with more instructions on top), which is why I decided to go for instruction count to distinguish between the two.

<details>
  <summary>18.48.39</summary>
  
image
image
  
</details>

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:92]: Done.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt (hunk 3 above)
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt:16]: Done.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 4]
```diff
@@ -65,34 +74,34 @@ object VideoInformationPatch : BytecodePatch(
...
 
             // add the seekTo method to the class for the integrations to call
             mutableClass.methods.add(seekHelperMethod)
         }
 
+        with(MdxPlayerDirectorFingerprint.result!!) {
+            mdxInitMethod = mutableClass.methods.first { MethodUtil.isConstructor(it) }
+
+            // find the location of the first invoke-direct call and extract the register storing the 'this' object reference
+            mdxInitInsertIndex = mdxInitMethod.implementation!!.instructions.indexOfFirst {
```
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt]: Done.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 5]
```diff
@@ -65,34 +74,34 @@ object VideoInformationPatch : BytecodePatch(
...
+                it.opcode == Opcode.INVOKE_DIRECT && ((it as Instruction35c).reference as MethodReference).name == "<init>"
```
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt]: Done.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt (hunk 2 above)
[the remark this answers is not in the document]
Comment (thedroidgeek) [reply src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt:7]: All tested.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 6]
```diff
@@ -65,34 +76,34 @@ object VideoInformationPatch : BytecodePatch(
...
+            mdxInitInsertIndex = mdxInitMethod.indexOfFirstInstructionOrThrow {
+                opcode == Opcode.INVOKE_DIRECT && getReference<MethodReference>()?.name == "<init>"
+            } + 1
+            mdxInitInsertRegister = mdxInitMethod.getInstruction<Instruction35c>(mdxInitInsertIndex - 1).registerC
+
+            // hook the MDX director for use through integrations
+            onCreateHookMdx(INTEGRATIONS_CLASS_DESCRIPTOR, "initializeMdx")
+
+            // MDX seek method
+            val mdxSeekFingerprintResultMethod = MdxSeekFingerprint.also { it.resolve(context, classDef) }.result!!.method
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:98]:
```suggestion
            val mdxSeekFingerprintResultMethod = MdxSeekFingerprint.apply { resolve(context, classDef) }.resultOrThrow().method
```

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 7]
```diff
@@ -158,6 +169,35 @@ object VideoInformationPatch : BytecodePatch(
...
+
+        // get enum type for the seek helper method
+        val seekSourceEnumType = seekMethod.parameterTypes[1].toString()
+
+        // insert helper method instructions
+        mdxSeekHelperMethod.addInstructions(
+            0,
+            """
+                sget-object v0, $seekSourceEnumType->a:$seekSourceEnumType
+                invoke-virtual {p0, p1, p2, v0}, ${seekMethod.definingClass}->${seekMethod.name}(J$seekSourceEnumType)Z
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:193]: Should work:

```suggestion
                invoke-virtual {p0, p1, p2, v0}, $seekMethod
```

Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:193]:
```suggestion
                invoke-virtual { p0, p1, p2, v0 }, ${seekMethod.definingClass}->${seekMethod.name}(J$seekSourceEnumType)Z
```

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt [hunk 8]
```diff
@@ -0,0 +1,18 @@
+package app.revanced.patches.youtube.video.information.fingerprints
+
+import app.revanced.patcher.extensions.or
+import app.revanced.patcher.fingerprint.MethodFingerprint
+import com.android.tools.smali.dexlib2.AccessFlags
+
+internal object MdxSeekFingerprint : MethodFingerprint(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt:7]: Quite the loose fingerprint, which could match other methods. Perhaps adding a small opcode pattern should make it a bit more failsafe.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt [hunk 9]
```diff
@@ -0,0 +1,8 @@
+package app.revanced.patches.youtube.video.information.fingerprints
+
+
+import app.revanced.patcher.fingerprint.MethodFingerprint
+
+internal object MdxPlayerDirectorFingerprint : MethodFingerprint(
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxPlayerDirectorFingerprint.kt:6]: Please name the fingerprint to the method it resolves to. If the name is obfuscated find a suitable name.

**Review by oSumAtrIX — CHANGES_REQUESTED** [review]

**Comment by thedroidgeek** [comment]
Sorry for the inactivity. @LisoUseInAIKyrios I just tested and everything seems to work properly with these changes.

**Review by thedroidgeek — COMMENTED** [review]
Review was done by @LisoUseInAIKyrios.

File: src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt [hunk 10]
```diff
@@ -42,57 +49,65 @@ object VideoInformationPatch : BytecodePatch(
...
+            val seekHelperMethod = generateSeekMethodHelper(seekFingerprintResultMethod)
 
             // add the seekTo method to the class for the integrations to call
             mutableClass.methods.add(seekHelperMethod)
         }
 
+        with(MdxPlayerDirectorSetVideoStageFingerprint.resultOrThrow()) {
+            mdxInitMethod = mutableClass.methods.first { MethodUtil.isConstructor(it) }
+
+            // find the location of the first invoke-direct call and extract the register storing the 'this' object reference.
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/VideoInformationPatch.kt:90]:
```suggestion
            // find the location of the first invoke-direct call and extract the register storing the 'this' object reference
```

File: src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt [hunk 11]
```diff
@@ -0,0 +1,23 @@
...
+    parameters = listOf("J", "L"),
+    opcodes = listOf(
+        Opcode.INVOKE_VIRTUAL,
+        Opcode.MOVE_RESULT,
+        Opcode.RETURN,
+    ),
+    customFingerprint = { methodDef, _ ->
+        // The instruction count is necessary here to avoid matching the relative version
+        // of the seek method we're after, which has the same function signature as the
+        // regular one, is in the same class, and even has the exact same 3 opcodes pattern
```
Comment (oSumAtrIX) [inline src/main/kotlin/app/revanced/patches/youtube/video/information/fingerprints/MdxSeekFingerprint.kt:20]:
```suggestion
        // regular one, is in the same class, and even has the exact same 3 opcodes pattern.
```

**Review by oSumAtrIX — APPROVED** [review]

**Comment by oSumAtrIX** [comment]
No

**Pull request merged**